avocado/utils/vmimage.py

```diff
@@ -0,0 +1,251 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: Red Hat Inc. 2017
+# Author: Amador Pahim <apahim@redhat.com>
+
+"""
+Provides VM images acquired from official repositories
+"""
+
+import os
+import tempfile
+import uuid
+
+from . import asset
+from . import path as utils_path
+from . import process
+
+try:
+    import lzma
+    LZMA_CAPABLE = True
+except ImportError:
+    LZMA_CAPABLE = False
+
+
+class VMImageError(Exception):
+    """
+    Generic error class for VM Image exceptions
+    """
+
+
+class Image(object):
+
+    def __init__(self, name, major, minor, build, arch, checksum, algorithm,
+                 cache_dir, url):
+        """
+        Vm Image representation.
+
+        :param name: Name of the Operating System
+        :param major: Major part of the version
+        :param minor: Minor part of the version
+        :param build: Build identifier
+        :param arch: Architecture
+        :param checksum: Hash of the image, used to validate the download
+        :param algorithm: Algorithm of the provided checksum
+        :param url: URL with the filename of the image
+        """
+        self.name = name
+        self.major = major
+        self.minor = minor
+        self.build = build
+        self.arch = arch
+        self.checksum = checksum
+        self.algorithm = algorithm
+        if cache_dir is None:
+            cache_dir = tempfile.gettempdir()
+        self.cache_dir = cache_dir
+        self.url = url
+        self.username = None
+        self.password = None
+        self.path = None
+        self.get()
+
+    def get(self):
+        """
+        Download the image with cache using the Asset Fetcher,
+        decompress image if needed and provides a backed image.
+        """
+        asset_path = asset.Asset(name=self.url,
+                                 asset_hash=self.checksum,
+                                 algorithm=self.algorithm,
+                                 locations=None,
+                                 cache_dirs=[self.cache_dir],
+                                 expire=None).fetch()
+
+        if os.path.splitext(asset_path)[1] == '.xz':
+            if LZMA_CAPABLE:
+                asset_path = self._extract(asset_path)
+            else:
+                raise VMImageError('Cannot uncompress image')
+
+        self.path = self._create_backed_image(asset_path)
+
+    def customize(self, username=None, password=None, install=None,
```

Comment:
I'd leave image customization out of the scene, initially.  The reason is that obtaining images is a problem in itself, and the customization is another one.  Depending on an external tool such as `virt-customize` is yet another hot topic.